pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,452 @@
+// Package libp2p implements the Mir transport interface using libp2p.
+package libp2p
+
+//go:generate go run ./gen/gen.go
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/filecoin-project/mir/pkg/events"
+	"github.com/filecoin-project/mir/pkg/logging"
+	mirnet "github.com/filecoin-project/mir/pkg/net"
+	"github.com/filecoin-project/mir/pkg/pb/eventpb"
+	"github.com/filecoin-project/mir/pkg/pb/messagepb"
+	"github.com/filecoin-project/mir/pkg/types"
+)
+
+const (
+	ProtocolID             = "/mir/0.0.1"
+	maxConnectingTimeout   = 1000 * time.Millisecond
+	maxRetryTimeout        = 2 * time.Second
+	maxRetries             = 20
+	noLoggingErrorAttempts = 2
+	PermanentAddrTTL       = math.MaxInt64 - iota
+)
+
+type TransportMessage struct {
+	Sender  string
+	Payload []byte
+}
+
+var _ mirnet.Transport = &Transport{}
+
+var ErrUnknownNode = errors.New("unknown node")
+var ErrNilStream = errors.New("stream has not been opened")
+
+type connInfo struct {
+	AddrInfo  *peer.AddrInfo
+	Stream    network.Stream
+	IsOpening bool
+}
+
+type Transport struct {
+	host             host.Host
+	ownID            types.NodeID
+	logger           logging.Logger
+	incomingMessages chan *events.EventList
+	connWg           *sync.WaitGroup
+	conns            map[types.NodeID]*connInfo
+	connsLock        sync.RWMutex
+}
+
+func NewTransport(h host.Host, ownID types.NodeID, logger logging.Logger) (*Transport, error) {
+	if logger == nil {
+		logger = logging.ConsoleErrorLogger
+	}
+
+	return &Transport{
+		ownID:            ownID,
+		host:             h,
+		incomingMessages: make(chan *events.EventList),
+		logger:           logger,
+		connWg:           &sync.WaitGroup{},
+		conns:            make(map[types.NodeID]*connInfo),
+	}, nil
+}
+
+func (t *Transport) Start() error {
+	t.logger.Log(logging.LevelDebug, "starting libp2p transport", "ownID", t.ownID, "listen", t.host.Addrs())
+	t.host.SetStreamHandler(ProtocolID, t.mirHandler)
+	return nil
+}
+
+func (t *Transport) Stop() {
+	t.logger.Log(logging.LevelDebug, "stopping libp2p transport", "ownID", t.ownID)
+	defer t.logger.Log(logging.LevelDebug, "stopping libp2p transport finished.", "ownID", t.ownID)
+
+	t.host.RemoveStreamHandler(ProtocolID)
+
+	t.connsLock.Lock()
+	for nodeID, addr := range t.conns {
+		if err := t.host.Network().ClosePeer(addr.AddrInfo.ID); err != nil {
+			t.logger.Log(logging.LevelError, "could not close old connection to node", "src", t.ownID, "dst", nodeID, "err", err)
+			continue
+		}
+	}
+	t.connsLock.Unlock()
```

Comment:
Does this prevent reconnection right afterwards?

Reply:
No, it doesn't. If you call `Stop` and then call `Connect` then new outbound connections can be established but new inbound connections to Mir protocol can't.
We can't close the `host` since we don't own it.

Reply:
I mean `Stop` and `reconnect` methods running concurrently.